### PR TITLE
API versioning mechanism and detection, message timestamps

### PR DESCRIPTION
**Pull request by dams**
This is the first part of more improvements to come.

In this PR I've implemented the ApiVersion endpoint, to know what the API versions the Kafka cluster supports. Based on that the Connection instance now maintains a map of supported API versions per endpoints.

I've then implemented v1, v2 and v3 versions of various API endpoints. I was specifically interested into TimeStamps supports in messages.

all tests pass on my box (including when specifying a kafka install dir), I hope it'll pass for everybody.

Next PR will be about supporting newer API versions regarding offsets queries and timestamps. Then I'll try to implement supprot for consumer groups.

Feedback welcome

### CODE REVIEW

File: lib/Kafka/Connection.pm [hunk 1]
```diff
@@ -498,6 +520,98 @@ sub get_known_servers {
...
+            my $implemented_max_version = $IMPLEMENTED_APIVERSIONS->{$api_key} // -1;
+            my $version = $kafka_max_version;
+            $version > $implemented_max_version
+              and $version = $implemented_max_version;
+            $version < $kafka_min_version
+              and $version = -1;
+            $self->{_api_versions}{$api_key} = $version;
+        }
+
+    };
```
Comment (asolovey) [inline lib/Kafka/Connection.pm]: There is no 'catch' for this try{}. I'd suggest catching errors and, depending on error, either throwing them or reporting them. See, for example, Kafka::Connection::_on_io_error.

**Comment by asolovey** [comment]
I run tests for this pull request locally, and a lot of them fail due to missing "offset_earliest" and "offset_latest" methods in Kafka::Consumer, for example:
```
t/20_kafka_usage.t .............. 1/? Can't locate object method "offset_latest" via package "Kafka::Consumer" at lib/Kafka/Consumer.pm line 585.
```
Also, one of the encode/decode tests fails as well:
```
t/06_decode_encode.t ............ 1/? 
#   Failed test 'decoded correctly'
#   at t/06_decode_encode.t line 707.
#     Structures begin differing at:
#          $got->{is_v1} = ''
#     $expected->{is_v1} = Does not exist
# Looks like you failed 1 test of 23.
```

File: t/config/server.properties [hunk 2]
```diff
@@ -20,7 +20,7 @@ broker.id=0
 # If this is set to true then attempts to produce, consume, or fetch metadata
 # for a non-existent topic will automatically create it
 # with the default replication factor and number of partitions.
-auto.create.topics.enable=false
+auto.create.topics.enable=true
```
Comment (asolovey) [inline t/config/server.properties]: I think this change is unintentional and should be reverted. Due to the way how cluster tests are implemented, this file is sometimes changed in the middle of the test and then reverted back. However, if the test is interrupted, the change remains and must be reverted manually.

Comment (dams) [inline t/config/server.properties]: you are right ! I'll fix it

File: lib/Kafka/Connection.pm (hunk 1 above)
Comment (dams) [inline lib/Kafka/Connection.pm]: Agreed. If the Kafka cluster version is < 0.10 ( I think), this API endpoint is not implemented, so a specific error is sent. I'll catch it explicitely and ignore it

**Comment by dams** [comment]
Ah, I have added more WIP commits to my master, after I had created the PR. So now this PR contains code that breaks. Sorry about that. I'll close this PR and open a new one.

There are some fix that I4m implementing, among them the fact that it's possible to detect the message format of a MessageSet, I thought it wasn't possible. So I don't need to keep the api_version around.

**Comment by dams** [comment]
closing, I'll reopen a new clean PR

**Pull request closed without merging**